Dockerfile

```diff
@@ -3,7 +3,11 @@ FROM node:13
 RUN mkdir -p /home/container
 WORKDIR /home/container
 
-COPY . /home/container
+# we should only re-run npm install if package.json specifically has changed. use docker build cache otherwise
+COPY ./package.json /home/container/package.json
 RUN npm i
 
+# copy everything else over
+COPY . /home/container
```

Comment:
???????
first you copy only package.json in this line you're copy all files??
I think that you can replace line 7 with
COPY . .   (you can do this because you have a dockerignore file with node_modules then you can copy all)
And delete this line

Reply:
The reason to do it this way is to differentiate between package.json changes and any other changes with docker build caching.

Check out the documentation here: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/

Each command creates a layer, and docker has certain caching strategies to determine if a layer needs to be rerun. The way I've written it, if you make a change to `package.json`, when you build a container, docker will detect that it needs to rerun from line:

`COPY ./package.json /home/container/package.json`

AKA, it will copy `package.json`, then install dependencies, and copy over all of the source code.

But, if you make changes to any other file, docker will detect that it can skip the `COPY ./package.json ...` command, will reuse the previously generated layer, and start from

`COPY . /home/container`

This is important because it will speed up the development flow when you are making changes to test files or other source files, but not touching any dependencies. It speeds up test run by at least a few seconds each time you run it.

Does this make sense?

Reply:
I will look over this soon!